### PR TITLE
kvs: fix memory use-after-free corner case

### PR DESCRIPTION
**Pull request by chu11**
My proposed change for #2500.  I hope I'm not posting this prematurely :-)  I was able to reproduce within unit tests once I figured out what was going on. Ran user reproducer and didn't get a segfault, although unclear if user's reproducer was successful, need to talk to user.

Basically, in the lookup handle I store a pointer to a treeobject (`wl->dirent`) which is owned by a cache entry.  I assume this pointer is valid, even after a lookup is stalled and restarted.

If the cache entry that owns the treeobj is destroyed via normal expiration methods -> lookup restarts after stall -> tries to read treeobj -> use-after-free.

In order to fix this, added some reference counters for cache entries, which I incref/decref as I save off pointers.

Went back and forth on multiple options, and this seemed to be the best.  It's admittedly more complex than I liked, but I think better than the alternatives.  It seemed wrong to do `json_incref/decref()` calls on const json objects.  Doing `json_deep_copy()` would have been a safe way to preserve const-ness, but some of these tree objects would be quite large (see #2505).  So I did it this way.

Fixed up a few stupid things along the way too.

### CODE REVIEW

**Comment by dongahn** [comment]
@chu11: Glad that you were able to come up with a patch so quickly. Once this lands, I will create a new version of flux for the user (fully load it with flux-sched and etc) and will ask @jameshcorbett to run the reproducer *and* if successful more production workloads. Since the MPI_Abort issue fix has already gone in, the new version should be much better for UQP.

**Comment by chu11** [comment]
> Ran user reproducer and didn't get a segfault, although unclear if user's reproducer was successful, need to talk to user.

@jameshcorbett said that things ran well, the issues in the run were due to the MPI abort issue.  He says he'll run some more.

**Comment by jameshcorbett** [comment]
Just to confirm what Al already found, I ran the ensemble again and everything looks good. I'll talk with David about coming up with some new tests.

File: src/modules/kvs/lookup.c [hunk 1]
```diff
@@ -459,8 +478,12 @@ static lookup_process_t walk (lookup_t *lh)
                 continue;
             }
         }
-        else
+        else {
+            cache_entry_decref (wl->entry);
             wl->dirent = dirent_tmp;
+            wl->entry = entry;
+            cache_entry_incref (wl->entry);
```
Comment (grondo) [inline src/modules/kvs/lookup.c]: Really minor suggestion -- I'm just noticing that this code is repeated at least a few times:

```C
cache_entry_decref (wl->entry);
wl->dirent = dirent_tmp;
wl->entry = entry;
cache_entry_incref (wl->entry);
```

Maybe an inline convenience function, like `walk_level_update (walk_level_t *wl, json_t *dirent, struct cache_entry *entry)`?

Comment (chu11) [reply src/modules/kvs/lookup.c]: yeah, it was just at that threshold of "should have I have a convenience function"?  Since you noticed it, we'll say it crosses the threshold now.

**Comment by dongahn** [comment]
@jameshcorbett: thank you for coming up with new tests. One thing that should be helpful for everyone is perhaps wait until this PR gets merged and I will install a new version. There are specific configurations that I would like you to test and I am willing to be a conf coondinator.

**Comment by jameshcorbett** [comment]
@dongahn Oh, of course. I just meant tests for Flux with the UQP in general, not this branch in particular.

**Comment by dongahn** [comment]
@jameshcorbett: thanks James. Yeah I thought this was your intent. I just wanted to make sure we are on the same page to coordinate-- getting the most bang for a buck!

**Comment by chu11** [comment]
re-pushed, adding the static inline convenience function per @grondo comments above

**Comment by garlick** [comment]
Great work!

**Comment by SteVwonder** [comment]
Awesome work everyone!  I've enjoyed reading through this debugging saga the past few days.

Are we waiting to close #2500 until @jameshcorbett has had a chance to re-run the ensemble cleanly with the latest fix?

**Comment by dongahn** [comment]
I just closed it. Thanks everyone!

**Comment by chu11** [comment]
> Are we waiting to close #2500 until @jameshcorbett has had a chance to re-run the ensemble cleanly with the latest fix?

Oh crud, in my commit message I closed the wrong issue

https://github.com/flux-framework/flux-core/commit/c82af7b238802c54ff300a7e797491cb34d29905

which I didn't notice b/c the first part of a fix issue for that was merged today.

**Comment by grondo** [comment]
Oh, sorry I didn't catch that! Bad reviewer!
